builtin/v13/migration/market.go

```diff
@@ -0,0 +1,503 @@
+		found, err := prevOutProviderSectors.Get(abi.UIntKey(uint64(miner)), &actorSectorsMapRoot)
+		if err != nil {
+			return cid.Undef, cid.Undef, xerrors.Errorf("failed to get actor sectors: %w", err)
+		}
+
+		var actorSectors *adt13.Map
+		if !found {
+			// can happen, this miner just didn't have any sectors before
+			actorSectors, err = adt13.MakeEmptyMap(ctxStore, market13.ProviderSectorsHamtBitwidth)
+			if err != nil {
+				return cid.Undef, cid.Undef, xerrors.Errorf("failed to make empty actor sectors map: %w", err)
+			}
+		} else {
+			actorSectors, err = adt13.AsMap(ctxStore, cid.Cid(actorSectorsMapRoot), market13.ProviderSectorsHamtBitwidth)
+			if err != nil {
+				return cid.Undef, cid.Undef, xerrors.Errorf("failed to load actor sectors map: %w", err)
+			}
+		}
+
+		for sector, deals := range sectors {
+			sectorDeals = deals
+
+			if err := actorSectors.Put(miner13.SectorKey(sector), &sectorDeals); err != nil {
+				return cid.Undef, cid.Undef, xerrors.Errorf("failed to put sector: %w", err)
+			}
+		}
+
+		newActorSectorsMapRoot, err := actorSectors.Root()
+		if err != nil {
+			return cid.Undef, cid.Undef, xerrors.Errorf("failed to get actor sectors root: %w", err)
+		}
+		actorSectorsMapRoot = typegen.CborCid(newActorSectorsMapRoot)
+
+		if err := prevOutProviderSectors.Put(abi.UIntKey(uint64(miner)), &actorSectorsMapRoot); err != nil {
+			return cid.Undef, cid.Undef, xerrors.Errorf("failed to put actor sectors: %w", err)
+		}
+	}
+
+	// get roots
+	outProviderSectors, err := prevOutProviderSectors.Root()
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to get providerSectorsRoot: %w", err)
+	}
+
+	outStates, err := prevOutStates.Root()
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to get statesRoot: %w", err)
+	}
+
+	return outProviderSectors, outStates, nil
+}
+
+func (m *marketMigrator) migrateProviderSectorsAndStatesFromScratch(ctx context.Context, store cbor.IpldStore, in migration.ActorMigrationInput, states cid.Cid) (cid.Cid, cid.Cid, error) {
+	ctxStore := adt.WrapStore(ctx, store)
+
+	oldStateArray, err := adt.AsArray(ctxStore, states, market12.StatesAmtBitwidth)
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load states array: %w", err)
+	}
+
+	newStateArray, err := adt13.MakeEmptyArray(ctxStore, market13.StatesAmtBitwidth)
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load states array: %w", err)
+	}
+
+	providerSectorsMem := map[abi.ActorID]map[abi.SectorNumber][]abi.DealID{}
+
+	var oldState market12.DealState
+	var newState market13.DealState
+	err = oldStateArray.ForEach(&oldState, func(i int64) error {
+		deal := abi.DealID(i)
+
+		newState.SlashEpoch = oldState.SlashEpoch
+		newState.LastUpdatedEpoch = oldState.LastUpdatedEpoch
+		newState.SectorStartEpoch = oldState.SectorStartEpoch
+		newState.SectorNumber = 0
+
+		if oldState.SectorStartEpoch != -1 {
```

Comment:
Can we do exactly what the FIP says to do:

- check whether `LastUpdatedEpoch` is -1
- if yes, find the sector in `providerSectors` (if you can't assert that the deal is past EndEpoch or fail)
- if no, just use 0

Reply:
> LastUpdatedEpoch is -1

`is` or `is not`? My belief is that LastUpdatedEpoch is only -1 before the deal is activated for the first time, so that sounds wrong.